capsules/src/lsm6dsoxtr.rs

```diff
@@ -0,0 +1,470 @@
+                            * (SCALE_FACTOR_GYRO[scale_factor] as isize)
+                            / 100) as usize;
+
+                        z = (((buffer[4] as u16 + ((buffer[5] as u16) << 8)) as i16) as isize
+                            * (SCALE_FACTOR_GYRO[scale_factor] as isize)
+                            / 100) as usize;
+                        nine_dof_client.callback(x, y, z)
+                    });
+                } else {
+                    self.nine_dof_client.map(|client| {
+                        client.callback(0, 0, 0);
+                    });
+                };
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.state.set(State::Idle);
+            }
+
+            State::ReadTemperature => {
+                let mut temperature: usize = 0;
+
+                if status == Ok(()) {
+                    self.temperature_client.map(|client| {
+                        temperature = ((((buffer[0] as u16 + ((buffer[1] as u16) << 8)) as i16)
+                            as isize
+                            / (TEMP_SENSITIVITY_FACTOR as isize)
+                            + 25)
+                            * 100) as usize;
+                        client.callback(temperature);
+                    });
+                } else {
+                    self.temperature_client.map(|client| {
+                        client.callback(0);
+                    });
+                };
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.state.set(State::Idle);
+            }
+
+            State::SetPowerModeAccel => {
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.state.set(State::Idle);
+                if status == Ok(()) {
+                    if self.config_in_progress.get() {
+                        if let Err(_error) = self.set_gyroscope_power_mode(
+                            self.gyro_data_rate.get(),
+                            self.low_power.get(),
+                        ) {
+                            self.config_in_progress.set(false);
+                        }
+                    }
+                }
+            }
+
+            State::SetPowerModeGyro => {
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.state.set(State::Idle);
+                self.config_in_progress.set(false);
+            }
+        }
+    }
+}
+
+impl SyscallDriver for Lsm6dsoxtrI2C<'_> {
+    fn command(
+        &self,
+        command_num: usize,
+        data1: usize,
+        data2: usize,
+        _process_id: ProcessId,
+    ) -> CommandReturn {
+        if command_num == 0 {
+            // Handle this first as it should be returned
+            // unconditionally
+            return CommandReturn::success();
+        }
+
+        match command_num {
```

Comment:
I'm concerned about these operations because they are split-phase but there is no resulting upcall. Userspace doesn't have a way to know when each operation, like setting the data rate/power mode, is finished.

At the very least there needs to be documentation on what each of these commands do and how the arguments should be passed. It's not really clear to me how important they are, as it would be nice to not have to special-case this particular accelerometer/magnetometer.